icu4c/source/common/locbased.cpp

```diff
@@ -11,85 +11,36 @@
-    const char* id = getLocaleID(valid, actual, type, status);
-    return Locale(id != nullptr ? id : "");
-}
-
-const char* LocaleBased::getLocaleID(const CharString* valid, const CharString* actual,
+const Locale& LocaleBased::getLocale(const Locale& valid, const Locale& actual,
                                      ULocDataLocaleType type, UErrorCode& status) {
     if (U_FAILURE(status)) {
-        return nullptr;
+        return Locale::getRoot();
     }
 
     switch(type) {
     case ULOC_VALID_LOCALE:
-        return valid == nullptr ? "" : valid->data();
+        return valid;
     case ULOC_ACTUAL_LOCALE:
-        return actual == nullptr ? "" : actual->data();
+        return actual;
     default:
         status = U_ILLEGAL_ARGUMENT_ERROR;
-        return nullptr;
+        return Locale::getRoot();
     }
 }
 
-void LocaleBased::setLocaleIDs(const CharString* validID, const CharString* actualID, UErrorCode& status) {
-    setValidLocaleID(validID, status);
-    setActualLocaleID(actualID,status);
-}
-void LocaleBased::setLocaleIDs(const char* validID, const char* actualID, UErrorCode& status) {
-    setValidLocaleID(validID, status);
-    setActualLocaleID(actualID,status);
-}
-
-void LocaleBased::setLocaleID(const char* id, CharString*& dest, UErrorCode& status) {
-    if (U_FAILURE(status)) { return; }
-    if (id == nullptr || *id == 0) {
-        delete dest;
-        dest = nullptr;
-    } else {
-        if (dest == nullptr) {
-            dest = new CharString(id, status);
-            if (dest == nullptr) {
-                status = U_MEMORY_ALLOCATION_ERROR;
-                return;
-            }
-        } else {
-            dest->copyFrom(id, status);
-        }
-    }
-}
+const char* LocaleBased::getLocaleID(const Locale& valid, const Locale& actual,
+                                     ULocDataLocaleType type, UErrorCode& status) {
+    const Locale& locale = getLocale(valid, actual, type, status);
 
-void LocaleBased::setLocaleID(const CharString* id, CharString*& dest, UErrorCode& status) {
-    if (U_FAILURE(status)) { return; }
-    if (id == nullptr || id->isEmpty()) {
-        delete dest;
-        dest = nullptr;
-    } else {
-        if (dest == nullptr) {
-            dest = new CharString(*id, status);
-            if (dest == nullptr) {
-                status = U_MEMORY_ALLOCATION_ERROR;
-                return;
-            }
-        } else {
-            dest->copyFrom(*id, status);
-        }
+    if (U_FAILURE(status)) {
+        return nullptr;
     }
-}
 
-bool LocaleBased::equalIDs(const CharString* left, const CharString* right) {
-    // true if both are nullptr
-    if (left == nullptr && right == nullptr) return true;
-    // false if only one is nullptr
-    if (left == nullptr || right == nullptr) return false;
-    return *left == *right;
+    return locale == Locale::getRoot() ? kRootLocaleName : locale.getName();
```

Comment:
kRootLocaleName is "root".
I believe that the current implementation just returns "", which I assume Locale::getRoot().getName() also returns, right?